hibernate-core/src/main/java/org/hibernate/dialect/PrestoDialect.java

```diff
@@ -0,0 +1,153 @@
+import java.sql.DriverPropertyInfo;
+import java.sql.SQLException;
+import java.sql.SQLFeatureNotSupportedException;
+import java.sql.Types;
+import java.util.Properties;
+import java.util.logging.Logger;
+
+import org.hibernate.boot.Metadata;
+import org.hibernate.dialect.function.VarArgsSQLFunction;
+import org.hibernate.dialect.pagination.AbstractLimitHandler;
+import org.hibernate.dialect.pagination.LimitHandler;
+import org.hibernate.dialect.pagination.LimitHelper;
+import org.hibernate.engine.spi.RowSelection;
+import org.hibernate.mapping.ForeignKey;
+import org.hibernate.tool.schema.spi.Exporter;
+import org.hibernate.type.StandardBasicTypes;
+
+/**
+ * Dialect for <a href="https://prestosql.io/">Presto</a>
+ *
+ * @author findepi
+ */
+public class PrestoDialect extends Dialect {
+
+	public PrestoDialect() {
+		registerHibernateType( Types.BIGINT, StandardBasicTypes.LONG.getName() );
+
+		registerFunction( "concat", new VarArgsSQLFunction( StandardBasicTypes.STRING, "", "||", "" ) );
+	}
+
+	@Override
+	public LimitHandler getLimitHandler() {
+		return new AbstractLimitHandler() {
+			@Override
+			public String processSql(String sql, RowSelection selection) {
+				final boolean hasOffset = LimitHelper.hasFirstRow( selection );
+				return sql + ( hasOffset
+						? " OFFSET ? LIMIT ?"
+						: " LIMIT ?" );
+			}
+
+			@Override
+			public boolean supportsLimit() {
+				return true;
+			}
+
+			@Override
+			public boolean supportsLimitOffset() {
+				return true;
+			}
+
+			@Override
+			public boolean supportsVariableLimit() {
+				return true;
+			}
+		};
+	}
+
+	@Override
+	public boolean dropConstraints() {
+		return false;
+	}
+
+	@Override
+	public Exporter<ForeignKey> getForeignKeyExporter() {
+		return new Exporter<ForeignKey>() {
+			@Override
+			public String[] getSqlCreateStrings(ForeignKey exportable, Metadata metadata) {
+				return new String[0];
+			}
+
+			@Override
+			public String[] getSqlDropStrings(ForeignKey exportable, Metadata metadata) {
+				return new String[0];
+			}
+		};
+	}
+
+	public static class PrestoDriverAdapter implements Driver {
+
+		private final Driver implementation;
+
+		public PrestoDriverAdapter() {
+			try {
+				implementation = (Driver) Class.forName( "io.prestosql.jdbc.PrestoDriver" ) // avoid compile-time dependency
+						.getConstructor()
+						.newInstance();
+			}
+			catch (ReflectiveOperationException e) {
+				throw new RuntimeException( e );
+			}
+		}
+
+		@Override
+		public Connection connect(String url, Properties info) throws SQLException {
+			if ( info != null ) {
+				info = copy( info );
+				// DdlTransactionIsolatorNonJtaImpl works with Presto driver only in autocommit mode (hibernate.connection.autocommit=true)
+				// When this is set, DriverConnectionCreator will pass autocommit connection property which Presto
+				// driver does not accept. Here we remove the property as a workaround, it's being handled on the
+				// Hibernate's side.
```

Comment:
I'm a bit concerned about this, could you expand on "it's being handled" ?

What are the implications of this, and what should users do when configuring a production connection pool and/or a Datasource configuration managed by an application server?

Reply:
Currently the `hibernate.connection.autocommit` mode is needed (at least for DDL in tests).
AFAIU this setting controls Hibernate behavior and also is passed to JDBC driver in the form of `autocommit` connection property.
The connection property can be safely ignored (Hibernate uses JDBC APIs to control autocommit of a connection), but Presto JDBC driver does not ignore connection properties that it does not understand.